### PR TITLE
Implement String#toUpperCase()

### PR DESCRIPTION
**Pull request by jtenner**
Alright. This is a functioning version of `String#toUpperCase()`.

After battling little endian (storing `u32` in two `u16` space) for a few days, I finally got it!

I'm sure we should implement a few lazy `Map` objects and had no good direction to go from here :).

Any advice?

### CODE REVIEW

**Comment by MaxGraey** [comment]
plz rebuild tests

**Comment by jtenner** [comment]
@MaxGraey Yeah I can rebuild the tests, but I need some help with optimization. Any advice?

**Comment by MaxGraey** [comment]
I guess will be great cover and pass all possible tests first before optimizaions)

**Comment by jtenner** [comment]
When I try to run the tests and create the test fixtures, for some reason I get the following errors.

```
  ERROR TS2339: Property 'toString' does not exist on type 'u64'.

     return arr[0].toString();
                   ~~~~~~~~
   in resolve-access.ts(3,16)

  ERROR TS2349: Cannot invoke an expression whose type lacks a call signature. Type 'resolve-access/Container#toU32' has no compatible call signatures.

     return container.toU32().toString();
            ~~~~~~~~~~~~~~~
   in resolve-access.ts(23,9)


  Error: Compile error
      at Object.main (C:\Users\jtenner\Desktop\projects\assemblyscript\cli\asc.js:492:21)
      at C:\Users\jtenner\Desktop\projects\assemblyscript\tests\compiler.js:231:9
      at Object.main (C:\Users\jtenner\Desktop\projects\assemblyscript\cli\asc.js:492:12)
      at C:\Users\jtenner\Desktop\projects\assemblyscript\tests\compiler.js:163:7
      at Array.forEach (<anonymous>)
      at Object.<anonymous> (C:\Users\jtenner\Desktop\projects\assemblyscript\tests\compiler.js:93:7)
      at Module._compile (internal/modules/cjs/loader.js:774:30)
      at Object.Module._extensions..js (internal/modules/cjs/loader.js:785:10)
      at Module.load (internal/modules/cjs/loader.js:641:32)
      at Function.Module._load (internal/modules/cjs/loader.js:556:12)
ERROR: 1 compiler tests had failures:

  resolve-access [Compile error]
```

**Comment by MaxGraey** [comment]
you should rebuild dist folder

**Comment by jtenner** [comment]
All set. Please review.

**Comment by MaxGraey** [comment]
Great!. Need much more tests. Try to find something in v8 or firefox tests. Or in tc39/test262
tests: https://github.com/tc39/test262/blob/ee3715ee56744ccc8aeb22a921f442e98090b3c1/test/built-ins/String/prototype/toUpperCase/special_casing.js

**Comment by MaxGraey** [comment]
Great! It will be great add also [this one](https://github.com/tc39/test262/blob/ee3715ee56744ccc8aeb22a921f442e98090b3c1/test/built-ins/String/prototype/toUpperCase/supplementary_plane.js)

File: tests/compiler/std/string.ts [hunk 1]
```diff
@@ -344,6 +344,150 @@ assert(dtoa(0.000035689) == "0.000035689");
 // assert(dtoa(f32.MAX_VALUE) == "3.4028234663852886e+38"); // FIXME
 // assert(dtoa(f32.EPSILON) == "1.1920928955078125e-7"); // FIXME
 
+
+// toUpperCase
+assert("abcdefghijklmnopqrstuvwxyz".toUpperCase() == "ABCDEFGHIJKLMNOPQRSTUVWXYZ", "-1");
+assert("aBcDeFgHiJkLmNoPqRsTuVwXyZ".toUpperCase() == "ABCDEFGHIJKLMNOPQRSTUVWXYZ", "-2");
+assert("AbCdEfGhIjKlMnOpQrStUvWxYz".toUpperCase() == "ABCDEFGHIJKLMNOPQRSTUVWXYZ", "-3");
+
+// Examples generated by @MaxGraey
```
Comment (MaxGraey) [inline tests/compiler/std/string.ts]: I guess this comment unnecessary)

**Comment by jtenner** [comment]
Thanks for showing me all the cases that needed to be covered. Looks like everything is all set. Submitting a commit now.

**Comment by MaxGraey** [comment]
Great. Also how do you think add some minor cases from `v8` engine?
https://github.com/v8/v8/blob/4b9b23521e6fd42373ebbcb20ebe03bf445494f9/test/intl/general/case-mapping.js#L36

File: tests/compiler/std/string.ts (hunk 1 above)
Comment (jtenner) [reply tests/compiler/std/string.ts]: I removed it, but it was just a comment so I could remember to visit them. If you have any more cases, please let me know. I will make this function as robust as I can.

```ts
store<u16>(out + outpos, 0xD801); // always store 0xD801
store<u16>(out + outpos + 2, char2 - 40);

// trying
store<u32>(out + outpos,
  (<u32>(char2 - 40) << <u32>16) & <u32>0xD801); // always store 0xD801(number - 40)
```

I'm going to try and turn this into a single store instruction, but this seems to error. Any help?

Comment (MaxGraey) [reply tests/compiler/std/string.ts]: It should be `((char2 - 40) << 16) | 0xD801`

Comment (jtenner) [reply tests/compiler/std/string.ts]: Never mind. It needs a `|` not an `&`.

**Comment by jtenner** [comment]
@MaxGraey I don't have any resources on these characters from the v8 engine.

```ts
// Plane 1; Deseret and Warang Citi Script.
// assert("\u{10400}\u{118A0}" == "\u{10428}\u{118C0}".toUpperCase(), "v8 test 21");
// assert("\u{1D400}\u{1D41A}" == "\u{1D400}\u{1D41A}".toUpperCase(), "v8 test 22");
// assert("\u{10C80}" == "\u{10CC0}".toUpperCase(), "v8 test 23")
```

These cases don't work because I don't account for these characters before-hand. 

All these plane 1 characters have special cases and I don't know how to test them.

**Comment by MaxGraey** [comment]
@dcodeIO it seems we haven't implementation for such string literals: `"\u{10C80}"`.

**Comment by jtenner** [comment]
Yeah I just need the set of plane 1 characters. Any advice on this?

**Comment by MaxGraey** [comment]
Just use `String.fromCodePoint(0x10C80)`

**Comment by jtenner** [comment]
Thanks! I found that I could inspect all the codepoints using a loop again.

At this point, we cover all test cases you have suggested so far except for a binary expression case created by `v8` that needs to be addressed by the resolver.

File: tests/compiler/std/string.ts [hunk 2]
```diff
@@ -527,6 +527,40 @@ assert("\uD801\uDC4D".toUpperCase() ==  "\uD801\uDC25",  "DESERET SMALL LETTER E
...
+assert("AΜ" == "aµ".toUpperCase(), "v8 test 13");
+assert("CSSBẶ" == "cßbặ".toUpperCase(), "v8 test 14");
+assert("FIFLFFIFFL" == "\uFB01\uFB02\uFB03\uFB04".toUpperCase(), "v8 test 15");
+assert("ABCÀCSSA" == "abcàcßa".toUpperCase(), "v8 test 16");
+assert("ABCDEFGHIÀCSSA" == "ABCDEFGHIàcßa".toUpperCase(), "v8 test 17");
+assert("ABCDEFGHIÀCSSA" == "abcdeFghiàcßa".toUpperCase(), "v8 test 18");
+
+// assert("ABCDE" == ("a" + "b" + "cde").toUpperCase(), "v8 test 19");
+assert("Α\u0301Ο\u0301Υ\u0301Ω\u0301" == "α\u0301ο\u0301υ\u0301ω\u0301".toUpperCase(), "v8 test 20");
+assert("\u{10400}\u{118A0}" == "\u{10428}\u{118C0}".toUpperCase(), "v8 test 21");
```
Comment (MaxGraey) [inline tests/compiler/std/string.ts]: wait! You still use it here =)

Comment (jtenner) [reply tests/compiler/std/string.ts]: What do you mean?

**Comment by jtenner** [comment]
I don't mind throwing more test cases and code planes at this. Let me know if I have missed something.

File: tests/compiler/std/string.ts (hunk 2 above)
Comment (MaxGraey) [reply tests/compiler/std/string.ts]: Is it supported ""\u{10400}\u{118A0}""?

Comment (jtenner) [reply tests/compiler/std/string.ts]: Just to make sure, this issue is addressed. All the plane 1 characters are supported now.

File: std/assembly/string.ts [hunk 3]
```diff
@@ -484,6 +484,1029 @@ import { idof } from "./builtins";
...
+      else if (char == 598) char = 393;
+      else if (char == 599) char = 394;
+      else if (char == 608) char = 403;
+      else if (char == 603) char = 400;
+      else if (char == 601) char = 399;
+
+      else if (char == 383
+        || char == 223
+        || char == 64261
+        || char == 64262) char = 83;
```
Comment (MaxGraey) [inline std/assembly/string.ts]: For `64261` and `64262` this not correct. They are return surrogate pair which only one code point contain `83`. I guess we have a lot of other similar issues. Plz recheck

**Pull request closed without merging**